### ISSUE TITLE
Task non-atomicity of interruption considered harmful

### ISSUE THREAD

**Issue by djspiewak** [functional-streams-for-scala/fs2#447]
This is something that we already decided to do away with in fs2 for simplicity and design reasons, but I think it's actually actively problematic even in the existing design.  Think about any sort of network IO or coordination problem, expressed as a series of composed `Task`(s), and then `eval`ed into scalaz-stream.  The coordination sequentialization and guarantees that you would expect *will not hold* simply because it's running within the `stepAsync` interpreter, meaning that the `Task` in question can be interrupted at any `bind` point without recourse or signal.

As it turns out, that's mildly inconvenient…  By show of hands, if I removed the `Task` cancelation semantics from 0.8 (in an 0.8.1 release, probably) such that any `Task` you `eval` (or equivalently, `await`) is guaranteed to either run fully to completion or *never* run at all, would anyone have any problems?  It would mean that you would need to be careful not to construct infinitely-processing tasks (e.g. `Process eval p1.repeat.runLog` for some non-empty process `p1` would be a no-no), but I think that, in general, infinities are idiomatically pushed into `Process` simply because they're easier to handle and schedule there.  So this shouldn't break anyone's *well constructed* code, and it should make some things (especially anything dealing with IO) much more reliable in the face of asynchronous interrupts and errors.

**Comment by pchlupacek** [comment]
@djspiewak I think we have to seriously consider that, even for fs2._

+1

**Comment by pchiusano** [comment]
Not allowing pre-emption mid task is I think the right decision but as for
whether we do a point release of 0.8 for it, I don't have a strong
preference and am fine with whatever. 0.9 will be out soon enough. :)

**Comment by pchlupacek** [comment]
@djspiewak going back to your original post I am not sure if I understand the intentions correctly

What would happen to task that is constructed like 
```scala

for {
 _ <- Task.async { cb =>  step1 }
 _ <- Task.fork { Task.delay { forked } } 
 _ <- Task.async { cb =>  step2 } 
} yield ???

```
When the resulting task  will receive interruption request between step1, exactly when `forked` is evaluated but before step2 was  evaluated?

**Comment by djspiewak** [comment]
@pchlupacek The interrupt flag is checked at bind points.  So if the interruption request shows up exactly as `forked` is evaluated, then chances are `forked` *will* run, but its result (if it completes) will be dropped on the floor.  `step2` will never run in this scenario.

@pchiusano Even with 0.9 coming out, we're not going to jump on the upgrade bandwagon that fast.  I hate to say it, but this is basically a "devil you know" scenario.  We'll upgrade to 0.9 once we've had a chance to test extensively with it, but given the massive, massive number of crazy and complicated cases that we've hit with our code base in 0.7 and in 0.8, I think we'll probably take it slow.

Hence the desire to do it in an 0.8.1.  :-)

**Comment by pchlupacek** [comment]
@djspiewak thanks. So the `Atomicity` task applies only for a `bind` scope. I was a bit curious if we are on same page how @pchiusano commented above where I was not sure if it applies to complete scope of the task. 

If, we however guarantee that bind is atomic in this sense as you suggest, I am still not grasping that full difference between atomic and non-atomic behaviour, hence, you still don't know where your `full` task was pre-empted.  What I mean, is if something got tossed away during the binds what is the chance that that would be exactly that `critical` piece which we really miss to perform IO operation/coordination correctly? I mean how you can be predictive here?

Just for curiosity, won't it be more practical to not have `bind` on the task altogether, and do just `bind` in process? In that case we can have `Task` fully atomic, and I think it may resolve other scenarios too. This of course only target fs2._. And I know this is very controversial (maybe even dumb:-)) question. But looking on our code, we do have much more `stream` flatMaps, and very, very very rare we use task flatMaps, that in fact always can be replaced with eval(task).map/flatMap. 

Another, approach to consider  (maybe for fs2._) is to remove `async` signature from Task alltogether. That will make Task interruption really unnecessary. And I would then leave async ONLY on Ref, and perhaps task._,  where we don't have flatMap (or have it on Stream._) , so if we need to interrupt we only interrupt that only and single `Get`, so we can be very, very atomic there.

**Comment by djspiewak** [comment]
@pchlupacek Removing `bind` from `Task` is an idea I raised to @pchiusano actually.  I still think it's a good one.  :-)  The problem with it *at the time* was that we were compiling `Process` into `Task` (via the interpreters), meaning that `Task` has to be at least as powerful as `Process`, which is to say a monad.  However, fs2 has now removed that design constraint, since a `Process[F, A]` is compiled into `Free[F, B]` (where `Monoid[B]`), meaning that `F` no longer needs to be a monad.  In other words, we could totally remove `bind` from `Task`.

In any case, interrupts *aren't* predictive, which is precisely the problem.  We can't predict where might be a good place to interrupt, so we just blindly interrupt at `bind` points.  Treating all tasks as uninterruptable gives full control to the user, since things don't randomly break "mid-stream" (which is exactly the problem that @alissapajer and I hit).

**Comment by pchiusano** [comment]
For FS2, I think Task must still be a monad. For one, I want an actual
runnable type in the library, rather than forcing people to bring their
own. Also, all the async functions (awaitAsync, etc), require an `Async`
instance which extends `Monad`, and the implementations make use of this.

I think it is still the right choice to make `Task` non-interruptible, and
use `flatMap` on `Stream` if you want to allow interruption. It is simple
to think about, and the implementation is easier too.

**Comment by pchlupacek** [comment]
@pchiusano can you please make clear to me what do you mean by making `Task non-interruptible` and how this can be achieved if there will be two Task.async bind over each other (not bringing the fact that this breaks trampoline too -> SOE issues ? I am not clear if I understand your definition of non-interruption clearly. 

Thanks.

**Comment by pchiusano** [comment]
I just mean that I don't want the library to provide a
"runAsyncInterruptibly" method or anything else that can prevent a Task
from running to completion once it starts running. If you want to stop a
Task, that is something you need to handle yourself, internal to the Task.
:)

Let me know if that makes sense.

**Comment by pchlupacek** [comment]
@pchiusano  I like what you did on `Async` trait, i.e. introduced pre-emption for any `F`, irrelevant how that will be implemented.

I think, there is perhaps a solution that won't sacrifice monadity, and perhaps leaves interruption logic to `F`, and still gives us deterministical behaviour of interruption. I would suggest:

- remove any runAsyncInterruptibly from the fs2.Task/Future
- keep Task.async and Future.async private
- introduce perhaps new combinators on task._ that will use `register` or `cb` kind of interface to current proposed ones, if necessary. 
- (cosmetic) In fact I would suggest task._ to be generalize to Async[F] and named it async._ dropping current queue, signal and topic from async package. 

The above guarantees that binding two async tasks would be possible only through Stream. I think this is in fact desired, and I don't see any case where this would limit usability // performance. 

What do you think? I think this will guarantee to us that interruption of the fs2.Task will be always atomic, and deterministic (run all or nothing).

**Comment by djspiewak** [comment]
@pchlupacek To pull back to the current master…  Making `Task` uninterruptable is as simple as commenting out the `set(false)` line in `completeInterruptibly` inside of `SourceSyntax`.  If that helps make things more concrete.

**Comment by pchiusano** [comment]
@pchlupacek those are interesting suggestions... let's discuss those separately from this ticket though, which Daniel originally opened regarding a 0.8 point release. Actually, let's discuss any FS2-related `Task` stuff on #441.

**Comment by pchlupacek** [comment]
@djspiewak yeah. for current master is simple and fully agree with that. @pchiusano sure. Sorry for cluttering this.

**Comment by pchiusano** [comment]
Closing this. 

All issues with interruption have been addressed for FS2 as of #526. If for some reason 0.9 drags on way too long and we decide we want to backport a change like this, we can reopen, but I'd like to say focused on getting 0.9 out the door.
